File: code/gaussian/generate_gaussian.R

# 1次元ガウス分布 --------------------------------------------------------------

# 1次元ガウス分布の生成


# パッケージの読込 ----------------------------------------------------------------

# 利用パッケージ
library(tidyverse)
library(cowplot)
library(magick)

# パッケージ名の省略用
library(ggplot2)


# ハイパーパラメータの影響 -----------------------------------------------------

### パラメータの設定 -----

# フレーム数を指定
frame_num <- 101

# 生成分布のパラメータを指定
mu_vals    <- seq(from = -5, to = 5, length.out = frame_num)
sigma_vals <- seq(from = 1, to = 1, length.out = frame_num)
#mu_vals    <- seq(from = 0, to = 0, length.out = frame_num)
#sigma_vals <- seq(from = 0, to = 10, length.out = frame_num+1)[-1] # (0を除去)

# サンプル分布のパラメータを指定
s <- 1


### 変数の設定 -----

# m軸の範囲を設定
u <- 5
sgm_num <- 1 # 倍率を指定
m_min <- (mu_vals - sgm_num*sigma_vals) |> # 基準値を指定
  min() |> 
  (\(.) {floor(. /u)*u})() # u単位で切り下げ
m_max <- (mu_vals + sgm_num*sigma_vals) |> # 基準値を指定
  max() |> 
  (\(.) {ceiling(. /u)*u})() # u単位で切り上げ
m_min; m_max

# m軸の値を作成
m_vec <- seq(from = m_min, to = m_max, length.out = 1001)
head(m_vec)


# x軸の範囲を設定
x_min <- m_min
x_max <- m_max
x_min; x_max

# x軸の値を作成
x_vec <- m_vec
head(x_vec)


### 対応関係の作図 -----

# 一時書き出し先を指定
dir_path <- "figure/tmp_folder"

# サンプルサイズを指定
N <- 10


# 確率密度軸の範囲を設定
u <- 0.05
gen_dens_max <- 1
gen_dens_max <- dnorm(
  x    = mu_vals, # 最頻値
  mean = mu_vals, 
  sd   = sigma_vals
) |> 
  max() |> 
  (\(.) {ceiling(. /u)*u})() # u単位で切り上げ
smp_dens_max <- dnorm(
  x    = mu_vals, # 最頻値
  mean = mu_vals, 
  sd   = s
) |> 
  max() |> 
  (\(.) {ceiling(. /u)*u})() # u単位で切り上げ
gen_dens_max; smp_dens_max

# ハイパラごとに作図
for(i in 1:frame_num) {
  
  ##### パラメータの生成 -----
  
  # 生成分布のパラメータを取得
  mu    <- mu_vals[i]
  sigma <- sigma_vals[i]
  
  # 期待値を計算
  E_m <- mu
  E_x <- E_m
  
  # サンプル分布のパラメータを生成
  pctl_min <- pnorm(q = m_min, mean = mu, sd = sigma)             # 描画範囲の最小値を取得
  pctl_max <- pnorm(q = m_max, mean = mu, sd = sigma)             # 描画範囲の最大値を取得
  pctl_n   <- seq(from = pctl_min, to = pctl_max, length.out = N) # 等間隔に累積確率を設定
  m_n      <- qnorm(p = pctl_n, mean = mu, sd = sigma)            # 分布の形状に応じて設定
  
  # 値を調整:(infの回避用)
  m_n[m_n < m_min] <- m_min
  m_n[m_n > m_max] <- m_max
  
  # サンプル分布のパラメータを格納
  smp_data_df <- tibble::tibble(
    n    = 1:N, # サンプル番号
    m    = m_n  # サンプル値
  )
  
  ##### 生成分布の作図 -----
  
  # 生成分布の確率密度を計算
  gen_dens_df <- tidyr::tibble(
    m    = m_vec, # 確率変数
    dens = dnorm(x = m, mean = mu, sd = sigma) # 確率密度
  )
  
  # 生成分布のラベルを作成
  gen_param_lbl <- paste0(
    "list(", 
    "mu == ", round(mu, digits = 2), ", ", 
    "sigma == ", round(sigma, digits = 2), 
    ")"
  ) |> 
    parse(text = _)
  gen_stats_lbl <- paste0(
    "list(", 
      "paste(", 
        "E(m) == mu, {} == ", round(E_m, digits = 2), 
      "), ", 
      "paste(", 
        "V(m) == sigma^2, {} == ", round(sigma^2, digits = 2), 
      ")", 
    ")"
  ) |> 
    parse(text = _)
  
  # 期待値のラベルを作成
  E_data_lbl  <- expression(E(m))
  
  # サンプルのラベルを作成
  smp_data_lbl <- paste0(
    "m[", 1:N, "]"
  ) |> 
    parse(text = _)
  smp_param_lbl <- paste0(
    "m[", 1:N, "] == ", round(m_n, digits = 2)
  ) |> 
    parse(text = _)
  
  # 生成分布を作図
  gen_graph <- ggplot() + 
    geom_vline(
      xintercept = E_m, 
      color = "red", linewidth = 1, linetype = "dashed"
    ) + # 期待値の位置
    geom_vline(
      data    = smp_data_df, 
      mapping = aes(xintercept = m, color = factor(n)), 
      linewidth = 1, linetype = "dotted", show.legend = FALSE
    ) + # サンプルの位置
    geom_line(
      data    = gen_dens_df, 
      mapping = aes(x = m, y = dens, linetype = "generator"), 
      linewidth = 1
    ) + # 生成分布
    geom_point(
      data    = smp_data_df, 
      mapping = aes(x = m, y = 0, color = factor(n)), 
      size = 4
    ) + # サンプル
    scale_x_continuous(
      sec.axis = sec_axis(
        trans  = ~ ., 
        breaks = c(E_m, m_n), 
        labels = c(E_data_lbl, smp_data_lbl), 
      ) # サンプルのラベル
    ) + 
    scale_color_hue(labels = smp_param_lbl) + # (凡例の表示用)
    scale_linetype_manual(
      breaks = "generator", 
      values = "solid", 
      labels = gen_param_lbl, 
      name   = "generator                 " # (スペースによる表示枠の固定)
    ) + # (凡例の表示用)
    guides(
      color    = "none", 
      #color    = guide_legend(override.aes = list(size = 4)), 
      linetype = guide_legend(override.aes = list(linewidth = 0.5))
    ) + 
    theme(
      axis.title    = element_text(size = 12), 
      plot.title    = element_text(size = 14), 
      plot.subtitle = element_text(size = 12)
    ) + # (グラフ位置のズレ対策用)
    coord_cartesian(
      xlim = c(m_min, m_max), 
      ylim = c(0, gen_dens_max)
    ) + # (軸の対応用)
    labs(
      title = "Gaussian distribution", 
      subtitle = gen_stats_lbl, 
      color = "sample", 
      x = expression(m), 
      y = expression(p(m ~"|"~ mu, sigma^2))
    )
  
  ##### サンプル分布の作図 -----
  
  # 期待値による分布の確率密度を計算
  E_dens_df <- tidyr::tibble(
    x    = x_vec, # 確率変数
    dens = dnorm(x = x, mean = E_m, sd = s) # 確率密度
  )
  
  # サンプル分布の確率密度を計算
  smp_dens_df <- tidyr::expand_grid(
    n = 1:N,   # サンプル番号
    x = x_vec, # 確率変数
  ) |> # サンプルごとに変数を複製
    dplyr::mutate(
      m    = m_n[n], # 平均パラメータ
      s    = s,      # 標準偏差パラメータ
      dens = dnorm(x = x, mean = m, sd = s) # 確率密度
    )
  
  # 期待値のラベルを作成
  E_data_lbl  <- expression(E(m))
  E_param_lbl <- paste0(
    "list(", 
    "E(m) == ", round(E_x, digits = 2), ", ", 
    "s == ", s, 
    ")"
  ) |> 
    parse(text = _)
  E_stats_lbl <- paste0(
    "list(", 
      "paste(", 
        "E(x) == mu, {} == ", round(E_x, digits = 2), 
        "), ", 
      "paste(", 
        "V(x) == s^2, {} == ", round(s^2, digits = 2), 
      ")", 
    ")"
  ) |> 
    parse(text = _)
  
  # サンプルのラベルを作成
  smp_data_lbl <- paste0(
    "m[", 1:N, "]"
  ) |> 
    parse(text = _)
  smp_param_lbl <- paste0(
    "list(", 
    "m[", 1:N, "] == ", round(m_n, digits = 2), ", ", 
    "s == ", s, 
    ")"
  ) |> 
    parse(text = _)
  
  # サンプル分布を作図
  smp_graph <- ggplot() + 
    geom_vline(
      xintercept = E_m, 
      color = "red", linewidth = 1, linetype ="dashed"
    ) + # 期待値との対応
    geom_vline(
      data    = smp_data_df, 
      mapping = aes(xintercept = m, color = factor(n)), 
      linewidth = 1, linetype = "dotted", show.legend = FALSE
    ) + # サンプルとの対応
    geom_line(
      data    = E_dens_df, 
      mapping = aes(x = x, y = dens, color = "expected"), 
      linewidth = 1, linetype = "dotdash"
    ) + # 期待値による分布
    geom_line(
      data    = smp_dens_df, 
      mapping = aes(x = x, y = dens, color = factor(n)), 
      linewidth = 1
    ) + # サンプルによる分布
    scale_x_continuous(
      sec.axis = sec_axis(
        trans  = ~ ., 
        breaks = c(E_m, m_n), 
        labels = c(E_data_lbl, smp_data_lbl), 
      ) # サンプルのラベル
    ) + 
    scale_color_manual(
      breaks = c("expected", 1:N), 
      values = c("red", scales::hue_pal()(n = N)), 
      labels = c(E_param_lbl, smp_param_lbl), 
      name   = "sample                    " # (スペースによる表示枠の固定)
    ) + # (凡例の表示用)
    guides(
      color = guide_legend(override.aes = list(linewidth = 0.5))
    ) + 
    theme(
      axis.title    = element_text(size = 12), 
      plot.title    = element_text(size = 14), 
      plot.subtitle = element_text(size = 12)
    ) + 
    coord_cartesian(
      xlim = c(x_min, x_max), 
      ylim = c(0, smp_dens_max)
    ) + # (軸の対応用)
    labs(
      title = "Gaussian distribution", 
      subtitle = E_stats_lbl, 
      x = expression(x), 
      y = expression(p(x ~"|"~ m, s^2))
    )
  
  ##### 対応関係の作図 -----
  
  # グラフを並べて描画
  comb_graph <- cowplot::plot_grid(
    gen_graph, smp_graph, 
    nrow = 2, ncol = 1, 
    align = "hv" # (グラフ位置のズレ対策用)
  )
  
  ##### グラフの出力 -----
  
  # 画像ファイルを書出
  file_path <- paste0(dir_path, "/", stringr::str_pad(i, width = nchar(frame_num), pad = "0"), ".png")
  ggplot2::ggsave(
    filename = file_path, plot = comb_graph, 
    width = 12, height = 9, units = "in", dpi = 100
  )
  
  # 途中経過を表示
  message("\r", i, " / ", frame_num, appendLF = FALSE)
}

# 動画を作成
paste0(dir_path, "/", stringr::str_pad(1:frame_num, width = nchar(frame_num), pad = "0"), ".png") |> # ファイルパスを作成
  magick::image_read() |> # pngファイルを読込
  magick::image_animate(fps = 1, dispose = "previous") |> # gifファイルを作成
  magick::image_write_video(path = "figure/gaussian/generate_gaussian/gauss_to_gauss.mp4", framerate = 30) -> tmp_path # mp4ファイルを書出

File: code/gaussian/generate_gaussian.py

# 1次元ガウス分布 ---------------------------------------------------------------

# 確率分布の生成
## 1次元ガウス分布の平均パラメータとの関係


# %%

# ライブラリの読込 ---------------------------------------------------------------

# ライブラリを読込
import numpy as np
from scipy.stats import norm
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation


# %%

# ハイパラとサンプル分布の形状の関係 ----------------------------------------------

### パラメータの設定 -----

# フレーム数を指定
frame_num = 101

# 生成分布のパラメータを指定
mu_vals    = np.linspace(start=-5.0, stop=5.0, num=frame_num)
sigma_vals = np.linspace(start=1.0, stop=1.0, num=frame_num)
#mu_vals    = np.linspace(start=0.0, stop=0.0, num=frame_num)
#sigma_vals = np.linspace(start=0.0, stop=10.0, num=frame_num+1)[1:] # (0を除去)
print('μ values:', mu_vals[:5])
print('σ values:', sigma_vals[:5])

# サンプル分布のパラメータを指定
s = 1.0


# %%

### 変数の設定 -----

# m軸の範囲を設定
u = 5.0
sgm_num = 1.0 # 倍率を指定
m_min = np.min(mu_vals - sgm_num*sigma_vals)
m_max = np.max(mu_vals + sgm_num*sigma_vals)
m_min = np.floor(m_min /u)*u # u単位で切り下げ
m_max = np.ceil(m_max /u)*u  # u単位で切り上げ
print('m-axis size:', m_min, m_max)

# m軸の値を作成
m_vec = np.linspace(start=m_min, stop=m_max, num=1001)
print('m values:', m_vec[:5])


# x軸の範囲を設定
x_min = m_min
x_max = m_max
print('x-axis size:', x_min, x_max)

# x軸の値を作成
x_vec = m_vec.copy()
print('x values:', x_vec[:5])


# %%

# サンプルサイズを指定
N = 7

# カラーマップを作成:(配色の共通化用)
cmap = plt.get_cmap('tab10') # カラーマップを指定
color_num = 10               # カラーマップの色数を設定


# 確率密度軸の範囲を設定
u = 0.05
gen_dens_max = np.max(norm.pdf(x=mu_vals, loc=mu_vals, scale=sigma_vals)) # 最頻値における値
gen_dens_max = np.ceil(gen_dens_max /u)*u # u単位で切り上げ
smp_dens_max = np.max(norm.pdf(x=mu_vals, loc=mu_vals, scale=s)) # 期待値による最頻値における値
smp_dens_max = np.ceil(smp_dens_max /u)*u # u単位で切り上げ
print('p-axis size:', gen_dens_max, smp_dens_max)

# 横軸の余白を追加
x_margin_rate = 0.05
u = 0.05
gen_x_margin_size = (m_max - m_min) * x_margin_rate
smp_x_margin_size = (x_max - x_min) * x_margin_rate
gen_x_margin_size = np.ceil(gen_x_margin_size /u)*u # u単位で切り上げ
smp_x_margin_size = np.ceil(smp_x_margin_size /u)*u # u単位で切り上げ
print('x-axis margin:', gen_x_margin_size, smp_x_margin_size)

# 縦軸の余白を追加
y_margin_rate = 0.049
u = 0.01
gen_y_margin_size = gen_dens_max * y_margin_rate
smp_y_margin_size = smp_dens_max * y_margin_rate
gen_y_margin_size = np.ceil(gen_y_margin_size /u)*u # u単位で切り上げ
smp_y_margin_size = np.ceil(smp_y_margin_size /u)*u # u単位で切り上げ
print('y-axis margin:', gen_y_margin_size, smp_y_margin_size)


# 図を初期化
fig, axes = plt.subplots(
    nrows=2, ncols=1, constrained_layout=True, 
    figsize=(10, 8), dpi=100, facecolor='white'
)
fig.suptitle('Gaussian distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    [ax.cla() for ax in axes]

    # パラメータを取得
    mu    = mu_vals[i]
    sigma = sigma_vals[i]

    # サンプルの期待値を計算
    E_m = mu
    E_x = E_m

    # パラメータを生成
    pctl_min = norm.cdf(x=m_min, loc=mu, scale=sigma)            # 描画範囲の最小値を取得
    pctl_max = norm.cdf(x=m_max, loc=mu, scale=sigma)            # 描画範囲の最大値を取得
    pctl_n   = np.linspace(start=pctl_min, stop=pctl_max, num=N) # 等間隔の累積確率に設定
    m_n      = norm.ppf(q=pctl_n, loc=mu, scale=sigma)           # 分布の形状に応じて設定

    # 値を調整:(infの回避用)
    m_n[m_n < m_min] = m_min
    m_n[m_n > m_max] = m_max

    # 確率分布を計算
    gen_dens_vec = norm.pdf(x=m_vec, loc=mu, scale=sigma) # 生成分布
    E_dens_vec   = norm.pdf(x=x_vec, loc=E_m, scale=s)    # 期待値による分布
    smp_dens_lt  = [norm.pdf(x=x_vec, loc=m_n[n], scale=s) for n in range(N)] # サンプル分布

    # 生成分布を描画
    ax = axes[0]
    ax.axvline(
        x=E_m, 
        color='red', linewidth=1.0, linestyle='--', 
        zorder=0
    ) # 期待値の位置
    ax.plot(
        m_vec, gen_dens_vec, 
        color='black', linewidth=1.0, 
        label=f'$\\mu = {mu:.2f}, \\sigma = {sigma:.2f}$', 
        zorder=2
    ) # 生成分布

    # サンプル分布を描画
    ax = axes[1]
    ax.axvline(
        x=E_x, 
        color='red', linewidth=1.0, linestyle='--', 
        zorder=0
    ) # 期待値の位置
    ax.plot(
        x_vec, E_dens_vec, 
        color='red', linewidth=1.0, linestyle='--', 
        zorder=2
    ) # 期待値による分布

    for n in range(N):
        
        # 値を取得
        m = m_n[n] # サンプル
        smp_dens_vec = smp_dens_lt[n] # 確率密度
        color_tp = cmap(n%color_num) # 色

        # 生成分布を描画
        ax = axes[0]
        ax.axvline(
            x=m, 
            color=color_tp, linewidth=1.0, linestyle=':', 
            zorder=1
        ) # サンプルの位置
        ax.scatter(
            x=m, y=0.0, 
            color=color_tp, s=50, 
            zorder=3
        ) # サンプル

        # サンプル分布を描画
        ax = axes[1]
        ax.axvline(
            x=m, 
            color=color_tp, linewidth=1.0, linestyle=':', 
            zorder=1
        ) # サンプルの位置
        ax.plot(
            x_vec, smp_dens_vec, 
            color=color_tp, linewidth=1.0, 
            label=f'$m_{{{n+1}}} = {m:.2f}, s = {s}$', 
            zorder=3
        ) # サンプル分布

    # 生成分布を装飾
    ax = axes[0]
    ax.set_xlabel('$m$') # m軸ラベル
    ax.set_ylabel('$p(m \\mid \\mu, \\sigma^2)$') # 確率密度軸ラベル
    ax.set_title(f'$E[m] = {E_m:.2f}$', loc='left') # 期待値ラベル
    ax.legend(title='gaussian', prop={'size': 8}, bbox_to_anchor=(1, 1), loc='upper left') # パラメータラベル
    ax.grid()
    ax.set_xlim(xmin=m_min-gen_x_margin_size, xmax=m_max+gen_x_margin_size) # (軸の対応用)
    ax.set_ylim(ymin=-gen_y_margin_size, ymax=gen_dens_max+gen_y_margin_size) # (軸の固定用)

    # サンプル分布を装飾
    ax = axes[1]
    ax.set_xlabel('$x$') # x軸ラベル
    ax.set_ylabel('$p(x \\mid m, s^2)$') # 確率密度軸ラベル
    ax.set_title(f'$E[x] = {E_x:.2f}$', loc='left') # 期待値ラベル
    ax.legend(title='gaussian', prop={'size': 8}, bbox_to_anchor=(1, 1), loc='upper left') # パラメータラベル
    ax.grid()
    ax.set_xlim(xmin=m_min-smp_x_margin_size, xmax=m_max+smp_x_margin_size) # (軸の対応用)
    ax.set_ylim(ymin=-smp_y_margin_size, ymax=smp_dens_max+smp_y_margin_size) # (軸の固定用)

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../figure/gaussian/generate_gaussian/gaussian_to_gaussian_mu.mp4', 
    progress_callback=lambda i, n: print(f'frame: {i} / {n}')
)


# %%
